lib/Lower/Bridge.cpp

```diff
@@ -574,41 +380,328 @@ class FirConverter : public Fortran::lower::AbstractConverter {
+      llvm::SmallVector<mlir::Value, 1> stepVector;
+      if (info.stepValue) {
+        stepVector.push_back(info.stepValue);
+      }
+      info.doLoop = builder->create<fir::LoopOp>(location, lowerValue,
+                                                 upperValue, stepVector);
+      builder->setInsertionPointToStart(info.doLoop.getBody());
+      mlir::Value shadow = builder->create<fir::ConvertOp>(
+          location, info.loopVariableType, info.doLoop.getInductionVar());
+      localSymbols.pushShadowSymbol(*info.loopVariableSym, shadow);
+      return;
+    }
 
-  void genFIR(const Fortran::parser::WhereConstruct &) { TODO(); }
+    // Unstructured loop preheader code - initialize tripVariable, loopVariable.
+    if (!info.stepValue) {
+      info.stepValue = builder->create<mlir::ConstantOp>(
+          location, builder->getIntegerAttr(info.loopVariableType, 1));
+    }
+    mlir::Value tripCount;
+    tripCount = builder->create<mlir::SubIOp>(location, upperValue, lowerValue);
+    tripCount =
+        builder->create<mlir::AddIOp>(location, tripCount, info.stepValue);
+    if (info.stepExpr.has_value()) {
+      tripCount = builder->create<mlir::SignedDivIOp>(location, tripCount,
+                                                      info.stepValue);
+    }
+    info.tripVariable = Fortran::lower::createTemporary(
+        location, *builder, localSymbols, info.loopVariableType, nullptr);
+    builder->create<fir::StoreOp>(location, tripCount, info.tripVariable);
+    builder->create<fir::StoreOp>(location, lowerValue, info.loopVariable);
+
+    // Unstructured loop header code - generate loop condition.
+    startBlock(info.headerBlock);
+    mlir::Value tripVariable =
+        builder->create<fir::LoadOp>(location, info.tripVariable);
+    mlir::Value zero = builder->create<mlir::ConstantOp>(
+        location, builder->getIntegerAttr(info.loopVariableType, 0));
+    mlir::Value cond = builder->create<mlir::CmpIOp>(
+        location, mlir::CmpIPredicate::sgt, tripVariable, zero);
+    genFIRConditionalBranch(cond, info.bodyBlock, info.successorBlock);
+  }
+
+  /// Generate FIR to end a structured or unstructured increment loop.
+  void genFIRIncrementLoopEnd(IncrementLoopInfo &info) {
+    mlir::Location location = toLocation();
+    if (info.isStructured()) {
+      // End fir.loop.
+      localSymbols.popShadowSymbol();
+      builder->restoreInsertionPoint(info.insertionPoint);
+      assert(info.doLoop.hasLastValue() && "loop must have a result");
+      mlir::Value result = builder->create<fir::ConvertOp>(
+          location, info.loopVariableType, info.doLoop.lastVal()[0]);
+      builder->create<fir::StoreOp>(location, result, info.loopVariable);
+      return;
+    }
+
+    // Unstructured loop - increment loopVariable.
+    mlir::Value loopVariable =
+        builder->create<fir::LoadOp>(location, info.loopVariable);
+    loopVariable =
+        builder->create<mlir::AddIOp>(location, loopVariable, info.stepValue);
```

Comment:
Stylistically, I prefer seeing each instruction get a variable name rather than reusing the name of another value. It makes it much easier to reorder and refactor code generation sequences.

Reply:
I can see advantages and disadvantages for both styles.  I'm inclined to have the option of using either one or the other, depending on circumstances.  Here I'm inclined to leave the code as is.  Note that this functionality is a candidate for future changes.  As is, the code uses a trip counter, because it works everywhere, including when the step is a runtime value.  (I've ignored non-integer loop bounds.  Those might eventually need extra work.)  But ideally, it might be good to generate either of the two more obvious loop conditions on the loop control variable up front, rather than counting on induction variable elimination to deal with those choices later.  But that can wait, possibly forever.